### PR TITLE
issue-12: implemented BinaryTree

### PR DESCRIPTION
**Pull request by kyparus**
Created AbstractTree and BinaryTree based on it.
Code is covered with tests.

### CODE REVIEW

File: src/core/trees/main/BinaryTree.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+     * Number of nodes in the tree
+     */
+    private var nodes_number = 0
+
+    /**
+     * Function to get number of nodes in the tree
+     */
+    override var size: Int = 0
+        get() = nodes_number
+
```
Comment (toniukan) [inline src/core/trees/main/BinaryTree.kt]: empty line

File: src/core/trees/main/BinaryTree.kt [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    /**
+     * Function to insert data to the tree
+     */
+    override fun insert(data: E) {
+        root = insert(root, data)
+        nodes_number++
+    }
+
+    /* Function to insert data recursively */
+    private fun insert(_node: BTNode<E>?, data: E): BTNode<E> {
```
Comment (toniukan) [inline src/core/trees/main/BinaryTree.kt]: It will build tree in this way. I think it's wrong.

photo4992334625137338293

Comment (toniukan) [inline src/core/trees/main/BinaryTree.kt]: maybe it should be "Binary Search Tree"?

Comment (kyparus) [inline src/core/trees/main/BinaryTree.kt]: Next will be BinarySearchTree

**Comment by Fly-Style** [comment]
Good job!

**Pull request merged**